File: python/problem_83.py
from typing import Optional, List


class TreeNode83:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right


def invert_tree(root: Optional[TreeNode83]) -> Optional[TreeNode83]:
    if root is None:
        return None
    root.left, root.right = invert_tree(root.right), invert_tree(root.left)
    return root


def inorder(root) -> List[int]:
    if root is None: return []
    return inorder(root.left) + [root.val] + inorder(root.right)


def main():
    print("Problem 83:")
    root = TreeNode83(4,
        TreeNode83(2, TreeNode83(1), TreeNode83(3)),
        TreeNode83(7, TreeNode83(6), TreeNode83(9))
    )
    assert inorder(invert_tree(root)) == [9, 7, 6, 4, 3, 2, 1]
    print("All Tests Passed")

File: typescript/problem_83.ts
class TreeNode83ts { constructor(public val:number, public left:TreeNode83ts|null=null, public right:TreeNode83ts|null=null){} }

export function invertTree(root: TreeNode83ts|null): TreeNode83ts|null {
    if (!root) return null;
    [root.left, root.right] = [invertTree(root.right), invertTree(root.left)];
    return root;
}

function inorder83ts(root: TreeNode83ts|null): number[] {
    if (!root) return [];
    return [...inorder83ts(root.left), root.val, ...inorder83ts(root.right)];
}

export function problem83(): void {
    console.log("Problem 83:");
    const root = new TreeNode83ts(4,
        new TreeNode83ts(2, new TreeNode83ts(1), new TreeNode83ts(3)),
        new TreeNode83ts(7, new TreeNode83ts(6), new TreeNode83ts(9))
    );
    if (JSON.stringify(inorder83ts(invertTree(root))) !== JSON.stringify([9,7,6,4,3,2,1]))
        throw new Error("assertion failed");
    console.log("All Tests Passed");
}
